design-proposals/host-ip-pool.md

```diff
@@ -0,0 +1,502 @@
+status:
+  allocated-offsets: [1, 2, 5, 6, 8]
+  conditions:
+  - type: Available
+    status: True
+    reason: ReconciledPool
+    message: The pool is up to date
+apiVersion: nmstate.io/v1
+kind: NodeIPPool
+metadata:
+  name: traffic-2
+spec:
+  range:
+    cidr: 10.10.11.0/24
+    start: 10.10.11.1
+    end: 10.10.10.11.10
+    exclude: ["10.10.11.4", "10.10.11.7"]
+status:
+  allocated-offsets: [2, 5, 6, 8]
+  conditions:
+  - type: Available
+    status: True
+    reason: ReconciledPool
+    message: The pool is up to date
+```
+
+The resource does not have a namespace since it's cluster wide information
+similar to NNCP. The `spec.range.cidr` is an IP range on CIDR format to
+configure IP pool, `spec.range.start` is the begining of the range and
+`spec.range.end` is the end, there is also an `spec.range.exclude` list of
+IPs or CIDR that will not be allocated.
+
+The `status.allocated-offsets` store the allocated IPs represented as a
+list of offsets from the range. Following are some example of possible pool
+configurations, allocated-offsets with their IP correlation:
+
+IPv4 /24 CIDR pool (submask 255.255.255.0)
+```yaml
+cidr: 10.10.11.0/24
+start: 10.10.11.4
+end: 10.10.10.11.10
+exclude: ["10.10.11.5", "10.10.11.7"]
+```
+
+offset -> IP:
+- 1 -> 10.10.11.4
+- 2 -> 10.10.11.5 (excluded)
+- 3 -> 10.10.11.6
+- 4 -> 10.10.11.7 (excluded)
+- 5 -> 10.10.11.8
+- 6 -> 10.10.11.9
+- 7 -> 10.10.11.10
+
+IPv4 /16 CIDR pool (submask 255.255.0.0)
+```yaml
+cidr: 10.10.11.0/16
+start: 10.10.11.4
+end: 10.10.10.12.10
+exclude: ["10.10.11.5", "10.10.12.7"]
+```
+
+offset -> IP:
+- 1 -> 10.10.11.4
+- 2 -> 10.10.11.5 (excluded)
+...
+- 251 -> 10.10.11.254
+- 252 -> 10.10.12.1
+- 253 -> 10.10.12.2
+...
+- 258 -> 10.10.12.7 (excluded)
+- 259 -> 10.10.12.8
+- 260 -> 10.10.12.9
+- 261 -> 10.10.12.10
+
+IPv4 /116 CIDR pool (4096 IPs)
+```yaml
+cidr: 2001:4860:4860::/116
+start: 2001:4860:4860::000F
+end: 2001:4860:4860::00FF
+exclude: ["2001:4860:4860::0010", "2001:4860:4860::001F"]
+```
+
+offset -> IP:
+- 1 -> 2001:4860:4860::000F
+- 2 -> 2001:4860:4860::0010 (excluded)
+...
+- 18 -> 2001:4860:4860::001F (excluded)
+- 19 -> 2001:4860:4860::0020
+...
+- 240 -> 2001:4860:4860::00FF
+
+Reseting `status.allocated-offset` field will trigger a Reconcile on the
+`NodeIPPool` and `allocated-offsets` will be re-calculated.
+In this process the reason will be "ReconcilingPool" and `Available`
+marked as unknown, so users can wait on it.
+
+Following is a table with the "type=Available" condition states
+
+| reason              | status  | message |
+| --------------------| --------| ------ |
+| ReconcilingPool     | Unknown | The pool is being updated |
+| ExhaustedIPPool     | False   | The IP pool has being depleted, manual intervention needed to modify NodeIPPool.|
+| UnsyncronizedIPPool | False   | Failed re constructing the IP pool, it may be not in sync with nodes. |
+| OverlapingIPPool    | False   | The NodeIPPool XYZ is overlapping with this one. |
```

Comment:
We shouldn't get this. But maybe it can be useful to have. `duplicate ip address`, in case two nodes got the same address somehow.

Reply:
I don't think that allocating a new address to one of them is a good option. Both may be running nodes, we cannot just change the IP of one of them.
The tool should prevent getting to this situation, but if the situation somehow happened (due to a bug), it should report about it.

Reply:
Fair enough, adding this error, we can implement some kind of remediation in the future, this can also come handy if users are configuring statically the cluster and they configured the same ip address at a pair of nodes by mistake.

Reply:
A user can be already assining the same address statically at the same interface, or at different interfaces, we can detect those.